engine/src/flutter/display_list/dl_paint.h

```diff
@@ -123,57 +123,77 @@ class DlPaint {
     return *this;
   }
 
-  std::shared_ptr<const DlColorSource> getColorSource() const {
+  const std::shared_ptr<const DlColorSource>& getColorSource() const {
     return color_source_;
   }
   const DlColorSource* getColorSourcePtr() const { return color_source_.get(); }
-  DlPaint& setColorSource(std::shared_ptr<const DlColorSource> source) {
-    color_source_ = std::move(source);
+
+  DlPaint& setColorSource(std::nullptr_t source) {
+    color_source_ = nullptr;
     return *this;
   }
   DlPaint& setColorSource(const DlColorSource* source) {
     color_source_ = source ? source->shared() : nullptr;
     return *this;
   }
+  DlPaint& setColorSource(std::shared_ptr<const DlColorSource> source) {
```

Comment:
What if this took a `shared_ptr&`?

Reply:
There's the same number of copies in the worst case, in the best case (where someone has an rvalue in setColorSource) it's less.

Reply:
(the current implementation)